src/internal/connector/exchange/service_iterators.go

```diff
@@ -50,6 +50,11 @@ func filterContainersAndFillCollections(
 		tombstones = makeTombstones(dps)
 	)
 
+	logger.Ctx(ctx).Infow(
+		"previous metadata",
+		"num_complete_items",
+		len(dps))
```

Comment:
Nit: messages should always refer to the current position in processing.  `logger.Ctx(ctx)` wraps `clues.In(ctx)`, so this log will naturally contain a lot more structured data than just the attached count.  That'll make it easy to lose the correlation between the log message and the structured data.
```suggestion
		"filling collections",
		"metadata_count",
		len(dps))
```